### DISCUSSION TITLE
Basic overriding configuration

### DISCUSSION THREAD

**Discussion by remy-poirier** [nvim-treesitter/nvim-treesitter#4216] [General]
### Describe the highlighting problem

I just would like to know how I can override the defaults highlighting. I'm new to VIM and I'm sorry to ask but I find that there is not that much of documentation about this.

In the snippet provided, I would like for example to highlight either the `import_clause` or the `import_specifier` to an other color, like Red for example, I tried some of the snippets that you'll find in the snippets section. I found them in the readme, but I cannot find a way to apply them to my configuration.

My configs are in lua

### Example snippet that causes the problem

I tried some snippets like :

```
vim.api.nvim_set_hl(0, '@named_imports.identifier', {
fg= '#00F00F', bg = '#FFFFFF'
})
```

```
vim.api.nvim_set_hl(0, '@import_statement.import_clause', {
fg= '#00F00F', bg = '#FFFFFF'
})
```

Unfortunately nothing has worked so far. However, it works for simple groups like `@comment` for example, the example below works:

```
vim.api.nvim_set_hl(0, "@comment", {
fg= '#00F00F', bg = '#FFFFFF'
})
```

I just would like to know the syntax I need to use to apply this to more precise Group.

Those lines are in the same file as the one where I setup the plugin, which looks like this

```
local status, ts = pcall(require, 'nvim-treesitter.configs')
if (not status) then return end

ts.setup {
  highlight = {
    enable = true,
    disable = {},
  },
  indent = {
    enable = true,
    disable = {}
  },
  ensure_installed = {
    'tsx',
    'lua',
    'json',
    'css',
    'html',
    'vim',
    'markdown'
  },
  autotag = {
    enable = true
  },
  query_linter = {
    enable = true,
    use_virtual_text = true,
    lint_events = {"BufWrite", "CursorHold"},
  },
}

vim.api.nvim_set_hl(0, "@interface_declaration.type_identifier", {
fg= '#00F00F'
})

vim.api.nvim_set_hl(0, "@comment", {
fg= '#00F00F', bg = '#FFFFFF'
})

vim.api.nvim_set_hl(0, '@named_imports.identifier', {
fg= '#00F00F', bg = '#FFFFFF'
})
```

### Tree-sitter parsing result

```
import_statement [0, 0] - [0, 80]
  import_clause [0, 7] - [0, 55]
    named_imports [0, 7] - [0, 55]
      import_specifier [0, 9] - [0, 16]
        name: identifier [0, 9] - [0, 16]
      import_specifier [0, 18] - [0, 22]
        name: identifier [0, 18] - [0, 22]
      import_specifier [0, 24] - [0, 34]
        name: identifier [0, 24] - [0, 34]
      import_specifier [0, 36] - [0, 41]
        name: identifier [0, 36] - [0, 41]
      import_specifier [0, 43] - [0, 53]
        name: identifier [0, 43] - [0, 53]
  source: string [0, 61] - [0, 80]
    string_fragment [0, 62] - [0, 79]
import_statement [1, 0] - [1, 30]
  import_clause [1, 7] - [1, 17]
    namespace_import [1, 7] - [1, 17]
      identifier [1, 12] - [1, 17]
  source: string [1, 23] - [1, 30]
    string_fragment [1, 24] - [1, 29]

```

### Example screenshot

I don't have any screenshots to provide as I'm not really able to achieve something :D 

### Expected behavior

I just would like to know how I can achieve a custom highlighting.

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - OK: `tree-sitter` found 0.20.7 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v18.13.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: Apple clang version 14.0.0 (clang-1400.0.29.202)
  - OK: Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

## OS Info:
{
  machine = "arm64",
  release = "21.6.0",
  sysname = "Darwin",
  version = "Darwin Kernel Version 21.6.0: Sat Jun 18 17:07:22 PDT 2022; root:xnu-8020.140.41~1/RELEASE_ARM64_T6000"
}

## Parser/Features         H L F I J
  - help                ✓ . . . ✓
  - html                ✓ ✓ ✓ ✓ ✓
  - tsx                 ✓ ✓ ✓ ✓ ✓
  - query               ✓ ✓ ✓ ✓ ✓
  - json                ✓ ✓ ✓ ✓ .
  - lua                 ✓ ✓ ✓ ✓ ✓
  - markdown            ✓ . ✓ . ✓
  - vim                 ✓ ✓ ✓ . ✓
  - c                   ✓ ✓ ✓ ✓ ✓
  - css                 ✓ . ✓ ✓ ✓

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}

```
```


### Output of `nvim --version`

```text
NVIM v0.8.2
Build type: Release
LuaJIT 2.1.0-beta3
Compilé par brew@HMBRW-A-001-M1-005.local

Features: +acl +iconv +tui
See ":help feature-compile"

         fichier vimrc système : "$VIM/sysinit.vim"
               $VIM par défaut : "/opt/homebrew/Cellar/neovim/0.8.2/share/nvim
"
```

### Additional context

I really hope someone will help me with this :)

**Comment by clason** [comment]
You need to assign highlight groups to the _captures_ (which are defined in the `highlights.scm` files), not the _nodes_ (which you are looking at in the parser output).

You should start with the captures in this list: https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights

**Reply by remy-poirier** [reply]
Hey,
Thank you for your quick answer, really ! 
So if I understand correctly, I can "only" customize the `captures` available in the link you provided ? I cannot have the granularity for example to provide a specific color for `import_statement -> import_clause -> named_imports -> import_specifier -> identifier` ? or at least something close to this ?

Also, do I have to create the `highlights.scm` file ?

**Reply by clason** [reply]
> So if I understand correctly, I can "only" customize the captures available in the link you provided ? 

No you can customize any capture you want; this list just gives you the ones that are used in nvim-treesitter.

> I cannot have the granularity for example to provide a specific color 

You cannot provide a color for nodes, period. But of course you can be very specific in capturing nodes with given capture, which you can then highlight.

> Also, do I have to create the highlights.scm file ?

Depends on what you want to do. Nvim-treesitter provides this file for all languages it can install; if you're not happy with those, you'll have to write your own.

I recommend reading the documentation `:h vim.treesitter` and this project's README from top to bottom.
